xgo.go

```diff
@@ -233,49 +233,58 @@ func pullDockerImage(image string) error {
 func compile(image string, config *ConfigFlags, flags *BuildFlags, folder string) error {
 	// If a local build was requested, find the import path and mount all GOPATH sources
 	locals, mounts, paths := []string{}, []string{}, []string{}
+	var usesModules bool
 	if strings.HasPrefix(config.Repository, string(filepath.Separator)) || strings.HasPrefix(config.Repository, ".") {
 		// Resolve the repository import path from the file path
 		config.Repository = resolveImportPath(config.Repository)
 
+		// Determine if this is a module-based repository
+		var modFile = config.Repository + "/go.mod"
```

Comment:
I recently discovered that `go env` provides `GOMOD` which makes this easier.

- When a repository contains a modules-based project, and `GO111MODULES` is either `on` or `auto`, `go env GOMOD` will contain the path to the `go.mod` file.
- When `GO111MODULES` is either `off` or `auto`, the `GOMOD` var will be the empty string.
- Interestingly, if `GO111MODULES` is set to `on` but the current repository is not a module-based project, then it will (at least sometimes) contain a path to a `go.mod` file which does not actually exist.

Reply:
>     * Interestingly, if `GO111MODULES` is set to `on` but the current repository is not a module-based project, then it will (at least sometimes) contain a path to a `go.mod` file which does not actually exist.

If thats the case I wouldn't want to rely on that env to find the mod file... Maybe this changes with Go 1.12 or 1.13, but as of now...

Reply:
[This same debate is being had by the `golang` team](https://github.com/golang/go/issues/29382) itself. It appears that `go env GOMOD` is intended to be The One Right Answer, but as the discussion on this issue illustrates, even the go build tool diverges. That same thread also confirms that [the algorithm for determining module status is volatile and likely to change](
https://github.com/golang/go/issues/29382#issuecomment-449482911) in the future.

Perhaps even more interesting is [this comment](
https://github.com/golang/go/issues/26500#issuecomment-407510940):

> I put GOMOD in 'go env' mainly so that it would be listed in bug reports. **The expected way to get at this is `go list -m -f {{.GoMod}}`** but I have not added that yet. :-)

The interesting thing is that, as of go 1.11.4 (at least), it actually *does* work (at least for my quick check on the command-line). So, worth looking at again in the future, especially given that `GOPATH` is expected to be deprecated at some point in the semi-distant future.

I'd like to see this PR get merged in the meanwhile--given how quickly people are adopting modules--and open issues + PR's to refactor and enhance later on.

Reply:
> I'd like to see this PR get merged in the meanwhile--given how quickly people are adopting modules--and open issues + PR's to refactor and enhance later on.

Yes, I'm with you on that.
Also, go modules will become the new standard in 1.13, which will first deprecate and later drop `GOPATH` entirely, so we should get this merged.

At Gitea, we also need this, and because this repo looks not maintained anymore [and other things](https://github.com/karalabe/xgo/issues/150) we may start a fork of xgo (and maintain it).

Reply:
@karalabe Can you chime in?